src/FakeItEasy/Creation/DelegateProxies/DelegateProxyGenerator.cs

```diff
@@ -74,21 +74,88 @@ private static Delegate CreateDelegateProxy(Type typeOfProxy, MethodInfo invokeM
         }
 
         private static Expression CreateBodyExpression(
-            MethodInfo delegateMethod, DelegateCallInterceptedEventRaiser eventRaiser, IEnumerable<Expression> parameterExpressions)
+            MethodInfo delegateMethod,
+            DelegateCallInterceptedEventRaiser eventRaiser,
+            ParameterExpression[] parameterExpressions)
         {
-            var parameterExpressionsCastToObject =
-                parameterExpressions.Select(x => Expression.Convert(x, typeof(object))).Cast<Expression>().ToArray();
-
-            Expression body = Expression.Call(
+            /*********************************************************************
+             * Generate a method that:
+             * - wraps its arguments in an object array
+             * - passes this array to eventRaiser.Raise()
+             * - assigns the output values back to the ref/out parameters
+             * - casts and returns the result of eventRaiser.Raise()
+             *
+             * For instance, for a delegate like this:
+             *
+             * delegate int Foo(int x, ref int y, out int z);
+             *
+             * We generate a method like this:
+             *
+             * int ProxyFoo(int x, ref int y, out int z)
+             * {
+             *     var arguments = new[]{ (object)x, (object)y, (object)z };
+             *     var result = (int)eventRaiser.Raise(arguments);
+             *     y = (int)arguments[1];
+             *     z = (int)arguments[2];
+             *     return result;
+             * }
+             ********************************************************************/
+
+            bool isVoid = delegateMethod.ReturnType == typeof(void);
+
+            // Local variables of the generated method
+            var arguments = Expression.Variable(typeof(object[]), "arguments");
+            var result = isVoid ? null : Expression.Variable(delegateMethod.ReturnType, "result");
+
+            var bodyExpressions = new List<Expression>();
+
+            // arguments = new[]{ (object)x, (object)y, (object)z };
+            var arrayInit =
+                Expression.NewArrayInit(
+                    typeof(object),
+                    parameterExpressions.Select(x => Expression.Convert(x, typeof(object))));
+            bodyExpressions.Add(Expression.Assign(arguments, arrayInit));
+
+            // eventRaiser.Raise(arguments)
+            Expression call = Expression.Call(
                 Expression.Constant(eventRaiser),
                 DelegateCallInterceptedEventRaiser.RaiseMethod,
-                new Expression[] { Expression.NewArrayInit(typeof(object), parameterExpressionsCastToObject) });
+                arguments);
+
+            if (!isVoid)
+            {
+                // If the return type is non void, cast the result of eventRaiser.Raise()
+                // to the real return type
+                call = Expression.Convert(call, delegateMethod.ReturnType);

```

Comment:
I would almost consider pulling `call = Expression.Assign(result, call)` from line 133 into this block, then having line 133 just becomes `bodyExpressions.Add(call)`, rather than having the reader re-evaluate the `isVoid`ness.